hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,251 @@
+    # doesn't even define this error. In this module, we simulate the Python 3.8+ behavior,
+    # defining and raising this error if necessary.
+    class InvalidStateError(Exception):
+        """Raised when attempting to change state of a future in a terminal state (e.g. finished)"""
 
-    TERMINAL_STATES = {base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED}
 
-    def __init__(self, connection: mp.connection.Connection):
-        """ manually create MPFuture. Please use MPFuture.make_pair instead """
+class UpdateType(Enum):
+    RESULT = auto()
+    EXCEPTION = auto()
+    CANCEL = auto()
+
+
+class MPFuture(base.Future, Generic[ResultType]):
+    """
+    A version of concurrent.futures.Future / asyncio.Future that can be fulfilled from a separate process.
+    Any process can access future status and set the result / exception and check for state.
+    However, only the original process (i.e. the process that created the future) can await the result or exception.
+
+    :note: This is an internal primitive that is not guaranteed to work outside of hivemind applications.
+     More specifically, there are two known limitations:
+       - MPFuture works between processes created through inheritance (e.g. fork), *not* for independent processes
+       - Different executors (non-origin processes) cannot call set_result / set_exception / cancel simultaneously
+    """
+    lock = mp.Lock()  # global lock that prevents simultaneous initialization and writing
+    pipe_waiter_thread: Optional[threading.Thread] = None  # process-specific thread that receives results/exceptions
+    global_mpfuture_receivers: Dict[PID, PipeEnd] = mp.Manager().dict()
+    active_futures: Optional[Dict[PID, MPFuture]] = None  # pending or running futures originated from current process
+    active_pid: Optional[PID] = None  # pid of currently active process; used to handle forks natively
+
+    def __init__(self, use_lock: bool = True,  loop: Optional[asyncio.BaseEventLoop] = None):
+        self._origin_pid, self._uid = os.getpid(), uuid.uuid4().int
+        self._shared_state_code = torch.empty([], dtype=torch.uint8).share_memory_()
+        self._state_cache = {}  # mapping from global to cached local future used that makes updates immediately
+        # available on setter side; dictionary-based cache works because future can visit any state at most once
+
+        base.Future.__init__(self)
         self._state, self._result, self._exception = base.PENDING, None, None
-        self.connection = connection
+        self._use_lock = use_lock
 
-    @classmethod
-    def make_pair(cls) -> Tuple[MPFuture, MPFuture]:
-        """ Create a pair of linked futures to be used in two processes """
-        connection1, connection2 = mp.Pipe()
-        return cls(connection1), cls(connection2)
+        if self.active_pid != self._origin_pid:
+            self._initialize_mpfuture_backend()
```

Comment:
Consider a process that didn't have MPFutures before. If two MPFutures are created simultaneously in this process, it will lead to a race condition in `self._initialize_mpfuture_backend()`.

I suggest to use a lock and the [double-checked locking pattern](https://en.wikipedia.org/wiki/Double-checked_locking) here.

Reply:
maybe it's easier to create a separate mp.Lock for initialization only?

Reply:
updated the code with double-checked locking
thanks!